### PR TITLE
Log telemetry event on container dispose

### PR DESCRIPTION
**Pull request by kian-thompson**
[How contribute to this repo](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md).

[Guidelines for Pull Requests](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

## Description

Disposing Container should log a telemetry event like closing a Container does. These events are critical for investigations.
See https://github.com/microsoft/FluidFramework/pull/13617

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -842,6 +842,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         try {
             // Ensure that we raise all key events even if one of these throws
             try {
+                // Raise event first, to ensure we capture _lifecycleState before transition.
+                // This gives us a chance to know what errors happened on open vs. on fully loaded container.
+                this.mc.logger.sendTelemetryEvent(
+                    {
+                        eventName: "ContainerDispose",
```
Comment (markfields) [inline packages/loader/container-loader/src/container.ts:849]: @vladsud - what do you think - "ContainerDispose", or "ContainerClose" with a prop saying it's dispose?  Or "ContainerClose_Dispose" :P

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -842,6 +842,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
+                        category: error === undefined ? "generic" : "error",
+                    },
+                    error,
+                );
+
                 // ! Progressing from "closed" to "disposing" is not allowed
                 if (this._lifecycleState !== "closed") {
```
Comment (markfields) [inline packages/loader/container-loader/src/container.ts:856]: @kian-thompson why is this necessary? Why not always transition to disposing?  They're equivalence-classed in this.closed right?

**Review by markfields — APPROVED** [review]

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:849]: > "ContainerClose" with a prop saying it's dispose

I vote for this.  Whether someone calls close or dispose should mostly be abstracted from us when we look at telemetry, it mostly means the same thing, and if we need to search out the distinction we'll have the easy way to query it.

Comment (markfields) [reply packages/loader/container-loader/src/container.ts:849]: Btw - The idea with "ContainerClose_Dispose" is we can update all our queries to `Data_eventName has "ContainerClose"` but the fact that it's Dispose is in your face right in the event name.  It's a little obscure/particular so probably not a good idea in practice.

File: packages/loader/container-loader/src/container.ts (hunk 2 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/container.ts:856]: This is a tricky nuance because the "closed" and "disposed" states are checked for the "Container.closed" boolean. In this new case, would we consider "disposing" to mean the Container is closed? My thinking was no, but we can't switch the state to non-closed as it may break a check somewhere.

File: packages/test/test-end-to-end-tests/src/test/container.spec.ts [hunk 3]
```diff
@@ -159,6 +160,7 @@ describeNoCompat("Container", (getTestObjectProvider) => {
     [
         { eventName: "fluid:telemetry:DeltaManager:GetDeltas_Exception", error: "expectedFailure" },
         { eventName: "fluid:telemetry:Container:ContainerClose", error: "expectedFailure" },
+        { eventName: "fluid:telemetry:Container:ContainerClose", error: "expectedFailure", disposed: true },
```
Comment (alexvy86) [inline packages/test/test-end-to-end-tests/src/test/container.spec.ts:163]: Does this mean we'll expect the container to raise two distinct ContainerClose events? Would it be better for the new one to be ContainerDisposed? I think our telemetry will get confusing otherwise.

Comment (kian-thompson) [reply packages/test/test-end-to-end-tests/src/test/container.spec.ts:163]: It's just so we align the naming. I agree that having a distinct event would be clean, but most people probably shouldn't have to know it can exist in place of ContainerClose.
The idea is that there can be situations where consumers of Container can just call "dispose", which technically also closes the container. So, we should log an event that is recognizable as a ContainerClose event without knowing these implementation details (could be done as "ContainerClose_Disposed" or with telemetry prop).

Comment (alexvy86) [reply packages/test/test-end-to-end-tests/src/test/container.spec.ts:163]: My concern is if calling dispose() generates _two_ ContainerClose events, one with no extra `disposed` prop and one with it. That's what I'd find confusing. Or if someone calls close first and later calls dispose once they've extracted everything they need from the container and are ready to fully get rid of it, that'd also result in two ContainerClose events, right?

Comment (kian-thompson) [reply packages/test/test-end-to-end-tests/src/test/container.spec.ts]: @alexvy86 Thanks for triggering this discussion. I brought it up with others and the consensus is to go with "ContainerDispose"

Comment (alexvy86) [reply packages/test/test-end-to-end-tests/src/test/container.spec.ts]: One further piece of feedback on that front: several of the tests that now expect the extra event are named in terms of "closing the container", e.g. "A large op will close the container". With the new distinction we're making, I think it's a good idea to update their names to be explicit about expectations of a container just closing VS a container closing and being disposed.
